src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -263,73 +372,141 @@ public String getRemote() {
 
-        listener.getLogger().format("%n  Getting remote branches...%n");
-        int branches = 0;
-        for (Map.Entry<String,GHBranch> entry : repo.getBranches().entrySet()) {
-            final String branchName = entry.getKey();
-            if (isExcluded(branchName)) {
-                continue;
-            }
-            listener.getLogger().format("%n    Checking branch %s%n", HyperlinkNote.encodeTo(repo.getHtmlUrl().toString() + "/tree/" + branchName, branchName));
-            if (criteria != null) {
-                SCMSourceCriteria.Probe probe = getProbe(branchName, "branch", "refs/heads/" + branchName, repo, listener);
-                if (criteria.isHead(probe, listener)) {
-                    listener.getLogger().format("    Met criteria%n");
-                } else {
-                    listener.getLogger().format("    Does not meet criteria%n");
+        // To implement buildOriginBranch && !buildOriginBranchWithPR we need to first find the pull requests, so we can skip corresponding origin branches later. Awkward.
+        Set<String> originBranchesWithPR = new HashSet<>();
+
+        if (buildOriginPRMerged || buildOriginPRUnmerged || buildForkPRMerged || buildForkPRUnmerged) {
+            listener.getLogger().format("%n  Getting remote pull requests...%n");
+            int pullrequests = 0;
+            for (GHPullRequest ghPullRequest : repo.getPullRequests(GHIssueState.OPEN)) {
+                int number = ghPullRequest.getNumber();
+                listener.getLogger().format("%n    Checking pull request %s%n", HyperlinkNote.encodeTo(ghPullRequest.getHtmlUrl().toString(), "#" + number));
+                boolean fork = !repo.getOwner().equals(ghPullRequest.getHead().getUser());
+                if (fork && !buildForkPRMerged && !buildForkPRUnmerged) {
+                    listener.getLogger().format("    Submitted from fork, skipping%n%n");
                     continue;
                 }
+                if (!fork && !buildOriginPRMerged && !buildOriginPRUnmerged) {
+                    listener.getLogger().format("    Submitted from origin repository, skipping%n%n");
+                    continue;
+                }
+                if (!fork) {
+                    originBranchesWithPR.add(ghPullRequest.getHead().getRef());
+                }
+                boolean trusted = isTrusted(repo, ghPullRequest);
+                if (!trusted) {
+                    listener.getLogger().format("    (not from a trusted source)%n");
+                }
+                for (boolean merge : new boolean[] {false, true}) {
+                    String branchName = "PR-" + number;
+                    if (merge && fork) {
+                        if (!buildForkPRMerged) {
+                            continue; // not doing this combination
+                        }
+                        if (buildForkPRUnmerged) {
+                            branchName += "-merged"; // make sure they are distinct
```

Comment:
`head` vs. `merge` would be OK too. I am not particularly moved by the argument of consistency with the terminology used in the API because (AFAICT) there is no mention in user documentation of virtual merge commits or the like.

“Head branch” vs. “base branch” _are_ used (briefly), so users could probably guess that `merge` refers to a merge of the two, whereas `head` does not.